common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java

```diff
@@ -904,13 +933,15 @@ public void onComplete(String streamId) throws IOException {
         // the error code to the client.
         AppShuffleMergePartitionsInfo info = appShuffleInfo.shuffles.get(partitionInfo.shuffleId);
         if (isTooLate(info, partitionInfo.reduceId)) {
-          deferredBufs = null;
+          freeAnyDeferredBlocks();
+          mergeManager.pushMergeMetrics.tooLateResponses.mark();
           throw new BlockPushNonFatalFailure(
             new BlockPushReturnCode(ReturnCode.TOO_LATE_BLOCK_PUSH.id(), streamId).toByteBuffer(),
             BlockPushNonFatalFailure.getErrorMsg(streamId, ReturnCode.TOO_LATE_BLOCK_PUSH));
         }
         if (isStale(info, partitionInfo.shuffleMergeId)) {
-          deferredBufs = null;
+          freeAnyDeferredBlocks();
+          mergeManager.pushMergeMetrics.noOpportunityResponses.mark();
```

Comment:
`couldNotFindOpportunityResponses` is to keep track of how many times a shuffle block collided because another block for the same reduce partition was being written. A stale block is different. It's from an executor that's from a previous app attempt/stage attempt. I don't think we should club them together. I don't think capturing how many stale blocks is that useful. 
cc. @Ngone51 @mridulm

Reply:
Fixed in the PR: https://github.com/apache/spark/pull/37638